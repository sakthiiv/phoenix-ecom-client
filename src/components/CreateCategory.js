import React, { Component } from "react";
import category from "../api/category";
import { Link } from "react-router-dom";

class CreateCategory extends React.Component {
  state = {
    cname: "",
    description: "",
    subcats: [{ name: "", description: "" }],
    registerStatus: false
  };
  handleChange = e => {
    if (["name"].includes(e.target.className)) {
      let subcats = [...this.state.subcats];
      subcats[e.target.dataset.id][e.target.className] = e.target.value;
      this.setState({ subcats });
    } else {
      this.setState({ [e.target.name]: e.target.value });
    }
  };
  addSubCat = e => {
    this.setState(prevState => ({
      subcats: [...prevState.subcats, { name: "" }]
    }));
  };
  handleSubmit = e => {
    e.preventDefault();
    var categoryObject = {
      description: this.state.description,
      name: this.state.cname,
      subCategory: this.state.subcats,
      isValid: "true"
    };
    category
      .create(categoryObject)
      .then(responseJson => {
        console.log(responseJson);
        this.setState({ registerStatus: true });
      })
      .catch(error => {
        console.log(error);
      });
  };
  render() {
    let { cname, description, subcats, registerStatus } = this.state;
    return (
      <div>
        {registerStatus ? (
          <div>
            <h1>Category created</h1>
            <Link to="/admin">Go to homepage</Link>
          </div>
        ) : (
          <div className="col-md-6">
            <h1>Create category</h1>
            <form onSubmit={this.handleSubmit} onChange={this.handleChange}>
              <div className="form-group">
                <label htmlFor="cname">Enter name of the category</label>
                <input
                  id="cname"
                  name="cname"
                  type="text"
                  placeholder="Enter name of the category"
                  className="form-control"
                  defaultValue={cname}
                  required
                />
              </div>

              <div className="form-group">
                <label htmlFor="description">
                  Example description of the category
                </label>
                <textarea
                  className="form-control"
                  name="description"
                  id="description"
                  rows="3"
                  defaultValue={description}
                  required
                />
              </div>
              <div className="form-group">
                <input
                  type="button"
                  value="Add subcategory"
                  className="add-sub-cat-btn btn-success"
                  onClick={this.addSubCat}
                />
              </div>
              {subcats.map((val, idx) => {
                let subCatId = `subcat-${idx}`;
                return (
                  <div key={idx}>
                    <div className="form-group sub-cat-item">
                      <label htmlFor={subCatId}>{`Sub Category  ${idx +
                        1}`}</label>
                      <input
                        type="text"
                        name={subCatId}
                        data-id={idx}
                        id={subCatId}
                        defaultValue={subcats[idx].name}
                        className="name"
                      />
                    </div>
                  </div>
                );
              })}
              <input type="submit" value="Submit" />
            </form>
          </div>
        )}
      </div>
    );
  }
}
export default CreateCategory;
