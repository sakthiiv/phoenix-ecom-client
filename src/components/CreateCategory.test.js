import React from 'react';
import { shallow } from 'enzyme';
import CreateCategory from './CreateCategory';
describe('CreateCategory', () => {
    it('renders correctly', () => {
      const wrapper = shallow(<CreateCategory />);
      // const onFormSubmit = { preventDefault: () => console.log('preventDefault') };
      // const onFormSubmit = jest.fn();
      const fakeEvent = { preventDefault: () => console.log('preventDefault') };
      expect(wrapper).toMatchSnapshot();
      // On the first run of this test, Jest will generate a snapshot file automatically.

      const formElement = wrapper.find('form'); // step 1 above
      formElement.simulate('submit', fakeEvent); // step 2
      expect(formElement.length).toBe(1);
      // expect(onFormSubmit).toHaveBeenCalledTimes(1); // step 3 
    });
});