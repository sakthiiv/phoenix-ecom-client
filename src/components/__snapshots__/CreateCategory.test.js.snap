// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCategory renders correctly 1`] = `
<div>
  <div
    className="col-md-6"
  >
    <h1>
      Create category
    </h1>
    <form
      onChange={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="cname"
        >
          Enter name of the category
        </label>
        <input
          className="form-control"
          defaultValue=""
          id="cname"
          name="cname"
          placeholder="Enter name of the category"
          required={true}
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="description"
        >
          Example description of the category
        </label>
        <textarea
          className="form-control"
          defaultValue=""
          id="description"
          name="description"
          required={true}
          rows="3"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="add-sub-cat-btn btn-success"
          onClick={[Function]}
          type="button"
          value="Add subcategory"
        />
      </div>
      <div
        key="0"
      >
        <div
          className="form-group sub-cat-item"
        >
          <label
            htmlFor="subcat-0"
          >
            Sub Category  1
          </label>
          <input
            className="name"
            data-id={0}
            defaultValue=""
            id="subcat-0"
            name="subcat-0"
            type="text"
          />
        </div>
      </div>
      <input
        type="submit"
        value="Submit"
      />
    </form>
  </div>
</div>
`;
